### PR TITLE
JPEG2000 drivers: do not export GMLJP2 if the SRS isn't compatible

### LINKED ISSUE #9223: JPEG2000 export: do not generate GMLJP2 for projections we don't support
GDAL OGRSpatialReference::exportToGML() is pretty limited and will for example generate incorrect content for
```
PROJCRS["Africa_Albers_Equal_Area_Conic",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["Albers Equal Area",
        METHOD["Albers Equal Area",
            ID["EPSG",9822]],
        PARAMETER["Latitude of false origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8821]],
        PARAMETER["Longitude of false origin",25,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8822]],
        PARAMETER["Latitude of 1st standard parallel",20,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8823]],
        PARAMETER["Latitude of 2nd standard parallel",-23,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8824]],
        PARAMETER["Easting at false origin",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8826]],
        PARAMETER["Northing at false origin",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8827]]],
    CS[Cartesian,2],
        AXIS["easting",east,
            ORDER[1],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]],
        AXIS["northing",north,
            ORDER[2],
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]]]
```
Not sure we want to invest in improving GML/XML export of SRS, but we should probably just error out / skip GMLJP2 if we can't write it correctly

### PR DESCRIPTION
**Pull request by rouault**
(fixes #9223)

### CODE REVIEW

File: autotest/gdrivers/ecw.py [hunk 1]
```diff
@@ -2210,6 +2210,66 @@ def test_ecw_read_uint32_jpeg2000():
...
+# Test unsupported XML SRS
+
+
+def test_jp2ecw_unsupported_srs_for_gmljp2(tmp_vsimem):
+
+    if gdaltest.jp2ecw_drv is None or not has_write_support():
+        pytest.skip("ECW write support not available")
+
+    filename = str(tmp_vsimem / "out.jp2")
+    # No EPSG code and Albers Equal Area not supported by OGRSpatialReference::exportToXML()
```
Comment (elpaso) [inline autotest/gdrivers/ecw.py:2223]: ?
```suggestion
    # No EPSG code and Albers Equal Area are supported by OGRSpatialReference::exportToXML()
```

File: autotest/gdrivers/jp2kak.py [hunk 2]
```diff
@@ -953,3 +953,60 @@ def find_xml_node(ar, element_name, only_attributes=False):
...
+
+
+###############################################################################
+# Test unsupported XML SRS
+
+
+def test_jp2kak_unsupported_srs_for_gmljp2(tmp_vsimem):
+
+    filename = str(tmp_vsimem / "out.jp2")
+    # No EPSG code and Albers Equal Area not supported by OGRSpatialReference::exportToXML()
```
Comment (elpaso) [inline autotest/gdrivers/jp2kak.py:965]: same as above ?

File: autotest/gdrivers/jp2openjpeg.py [hunk 3]
```diff
@@ -3864,3 +3864,62 @@ def test_jp2openjpeg_limit_resolution_count_from_image_size(tmp_vsimem):
...
+
+
+###############################################################################
+# Test unsupported XML SRS
+
+
+def test_jp2openjpeg_unsupported_srs_for_gmljp2(tmp_vsimem):
+
+    filename = str(tmp_vsimem / "out.jp2")
+    # No EPSG code and Albers Equal Area not supported by OGRSpatialReference::exportToXML()
```
Comment (elpaso) [inline autotest/gdrivers/jp2openjpeg.py:3876]: same ?

File: frmts/ecw/ecwcreatecopy.cpp [hunk 4]
```diff
@@ -1032,10 +1032,33 @@ CPLErr GDALECWCompressor::Initialize(
...
                     CSLFetchNameValue(papszOptions, "GMLJP2V2_DEF");
                 if (pszGMLJP2V2Def != nullptr)
+                {
                     WriteJP2Box(oJP2MD.CreateGMLJP2V2(nXSize, nYSize,
                                                       pszGMLJP2V2Def, poSrcDS));
+                }
                 else
-                    WriteJP2Box(oJP2MD.CreateGMLJP2(nXSize, nYSize));
+                {
+                    if (poSRS == nullptr || poSRS->IsEmpty() ||
```
Comment (elpaso) [inline frmts/ecw/ecwcreatecopy.cpp:1041]:
```suggestion
                    if (!poSRS || poSRS->IsEmpty() ||
```

File: frmts/jp2kak/jp2kakdataset.cpp [hunk 5]
```diff
@@ -2701,26 +2701,50 @@ static GDALDataset *JP2KAKCreateCopy(const char *pszFilename,
...
                 poBox = oJP2MD.CreateGMLJP2V2(nXSize, nYSize, pszGMLJP2V2Def,
                                               poSrcDS);
             }
             else
             {
-                poBox = oJP2MD.CreateGMLJP2(nXSize, nYSize);
+                const OGRSpatialReference *poSRS =
+                    poSrcDS->GetGCPCount() > 0 ? poSrcDS->GetGCPSpatialRef()
+                                               : poSrcDS->GetSpatialRef();
+                if (poSRS == nullptr || poSRS->IsEmpty() ||
```
Comment (elpaso) [inline frmts/jp2kak/jp2kakdataset.cpp:2715]:
```suggestion
                if (!poSRS || poSRS->IsEmpty() ||
```

File: frmts/jp2lura/jp2luradataset.cpp [hunk 6]
```diff
@@ -713,10 +713,19 @@ GDALDataset *JP2LuraDataset::CreateCopy(const char *pszFilename,
             {
                 bGeoreferencingCompatOfGeoJP2 = true;
                 oJP2MD.SetGeoTransform(adfGeoTransform);
+
+                if (poSRS != nullptr && !poSRS->IsEmpty())
```
Comment (elpaso) [inline frmts/jp2lura/jp2luradataset.cpp:717]:
```suggestion
                if (poSRS && !poSRS->IsEmpty())
```

File: frmts/opjlike/jp2opjlikedataset.cpp [hunk 7]
```diff
@@ -2471,10 +2471,18 @@ GDALDataset *JP2OPJLikeDataset<CODEC, BASE>::CreateCopy(
             {
                 bGeoreferencingCompatOfGeoJP2 = TRUE;
                 oJP2MD.SetGeoTransform(adfGeoTransform);
+                if (poSRS != nullptr && !poSRS->IsEmpty())
```
Comment (elpaso) [inline frmts/opjlike/jp2opjlikedataset.cpp:2474]:
```suggestion
                if (poSRS && !poSRS->IsEmpty())
```

**Review by elpaso — APPROVED** [review]

File: autotest/gdrivers/ecw.py (hunk 1 above)
Comment (rouault) [reply autotest/gdrivers/ecw.py]: changed to "    # There is no EPSG code and Albers Equal Area is not supported by OGRSpatialReference::exportToXML()"
done

File: autotest/gdrivers/jp2kak.py (hunk 2 above)
Comment (rouault) [reply autotest/gdrivers/jp2kak.py]: done

File: autotest/gdrivers/jp2openjpeg.py (hunk 3 above)
Comment (rouault) [reply autotest/gdrivers/jp2openjpeg.py]: done

File: frmts/ecw/ecwcreatecopy.cpp (hunk 4 above)
Comment (rouault) [reply frmts/ecw/ecwcreatecopy.cpp]: done

File: frmts/jp2kak/jp2kakdataset.cpp (hunk 5 above)
Comment (rouault) [reply frmts/jp2kak/jp2kakdataset.cpp]: done

File: frmts/jp2lura/jp2luradataset.cpp (hunk 6 above)
Comment (rouault) [reply frmts/jp2lura/jp2luradataset.cpp]: done

File: frmts/opjlike/jp2opjlikedataset.cpp (hunk 7 above)
Comment (rouault) [reply frmts/opjlike/jp2opjlikedataset.cpp]: done

**Pull request merged**